### PR TITLE
DDF-3636: Decouple Applications from karaf features

### PR DESCRIPTION
**Pull request by tbatie**
#### What does this PR do?
Decouples applications from karaf. Applications only serve the purpose of categorizing configurations in the admin console with these changes. The application features still exist and can be started from the system tab.

Backend changes:
- Deprecated application interfaces
- Remove app state and functionality including the install, uninstall, start and stop. ()
- Moved app definitions to configuration filese
- Fixed circular dependencies in feature files so app features can start up properly.

Frontend changes:
- Only allow profiles to be selected during installation.
- Removed dead code related to managing app state
- Removed features & details tabs of applications

#### Who is reviewing it? 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
@ryeats 
@emanns95 
@paouelle 

#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@figliold 
@pklinef 

#### How should this be tested? (List steps with links to updated documentation)
Ensure install profiles are working as expected and each application tab shows configurations and plugins as before.

#### Screenshots (if appropriate)
screen shot 2018-02-21 at 12 35 07 pm
screen shot 2018-02-21 at 12 34 58 pm

#### Checklist:
- [x] Documentation Updated (Will updated in separate pr)
- [x] Update / Add Unit Tests
- [x] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by tbatie** [comment]
build now

File: platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 1]
```diff
@@ -11,20 +11,15 @@
...
 
-    <bean id="appService"
-          class="org.codice.ddf.admin.application.service.impl.ApplicationServiceImpl">
+    <bean id="appService" class="org.codice.ddf.admin.application.service.impl.ApplicationServiceImpl">
         <cm:managed-properties
             persistent-id="org.codice.ddf.admin.application.service.impl.ApplicationServiceImpl"
             update-strategy="container-managed"/>
-        <argument ref="bundleStateServices"/>
-        <property name="configFileName"
-                  value="${ddf.home}/etc/org.codice.ddf.admin.applicationlist.properties"/>
```
Comment (ryeats) [inline platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: ❓Since your getting rid of applicationlist.properties can we now accomplish the same thing by making our apps boot features?

File: platform/security/security-services-app/src/main/resources/features.xml [hunk 2]
```diff
@@ -331,4 +325,9 @@
         <feature>security-sts-realm</feature>
     </feature>
 
+    <feature name="security-services-app" install="auto" version="${project.version}"
+             description="The Security Application provides Authentication, Authorization, and Auditing services.\nThey comprise both a framework that developers and integrators can extend and also a reference implementation that can be used which meets security requirements.::Security">
```
Comment (ryeats) [inline platform/security/security-services-app/src/main/resources/features.xml]: ✏️can we get rid of the `::Security` now? same comment for all the apps really.

File: distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "admin-app",
+  "description": "Administration application for installation and management.\nIncludes the Admin UI and the underlying application service that supports the user interface.\nThe Admin UI provides contains modules allowing the administrator to access configuration pages to tailor system services and properties.::Admin",
```
Comment (ryeats) [inline distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json]: ✏️can we get away from encoding the app name in the description i.e. `::Admin`

File: platform/security/security-services-app/src/main/resources/features.xml (hunk 2 above)
Comment (tbatie) [reply platform/security/security-services-app/src/main/resources/features.xml]: Although the backend no longer uses the end of the description, the frontend is parsing it to give it a display name. I originally removed them but the app names were no longer user friendly. I was going to add a displayName field and method to the Application interface until I started deprecating it.

File: platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 1 above)
Comment (tbatie) [reply platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: Yeah, you should be able to now since all the bundles to display in the admin console are now defined in the app feature. Although, I question how stable it would be to start up app features independently right now. There's still a lot of missing dependencies in app features that are only working because of the profiles starting up more than one app.

File: distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json (hunk 3 above)
Comment (tbatie) [reply distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json]: Same comment

File: platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 1 above)
Comment (ryeats) [reply platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: hmm, can we leave this functionality in then? or will another PR be addressing the dependancies shortly? I would like to still be able to skip the install screen and just start all the features my distribution needs on startup.

File: platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceImpl.java [hunk 4]
```diff
@@ -463,427 +153,48 @@ public Application findFeature(Feature feature) {
     List<Feature> profiles = new ArrayList<>();
     try {
       profiles =
-          Arrays.stream(featuresService.listFeatures())
+          Arrays.asList(featuresService.listFeatures())
```
Comment (paouelle) [inline platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceImpl.java]: ❓ Why the change? This now creates a temporary collection just to stream it.

File: platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceImpl.java [hunk 5]
```diff
@@ -463,427 +153,48 @@ public Application findFeature(Feature feature) {
...
-          feature.getVersion(),
-          e);
-    }
-  }
+  public void removeApplication(String applicationName) throws ApplicationServiceException {}
 
   @Override
   public List<FeatureDetails> getAllFeatures() {
-    List<FeatureDetails> features = new ArrayList<>();
+    List<FeatureDetails> features = new ArrayList<FeatureDetails>();
```
Comment (paouelle) [inline platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceImpl.java]: ✏️ You should not add the type in between `<>`. Static analysis will complain.

File: platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceImpl.java [hunk 6]
```diff
@@ -897,7 +208,7 @@ private void tryUninstallFeature(Feature feature) {
   }
 
   private Map<String, String> getFeatureToRepository() {
-    Map<String, String> feature2repo = new HashMap<>();
+    Map<String, String> feature2repo = new HashMap<String, String>();
```
Comment (paouelle) [inline platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceImpl.java]: Same comment as above

File: platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/command/AbstractProfileCommandTest.java [hunk 7]
```diff
@@ -1,104 +0,0 @@
-/**
```
Comment (paouelle) [inline platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/command/AbstractProfileCommandTest.java]: ❗️ I don't think we can remove this one. It is required for the profile install and profile list commands.

File: platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/command/ProfileInstallCommandTest.java [hunk 8]
```diff
@@ -1,349 +0,0 @@
-/**
- * Copyright (c) Codice Foundation
```
Comment (paouelle) [inline platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/command/ProfileInstallCommandTest.java]: ❗️ We should not be removing this command

File: platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/command/ProfileListCommandTest.java [hunk 9]
```diff
@@ -1,149 +0,0 @@
-/**
```
Comment (paouelle) [inline platform/admin/core/admin-core-appservice/src/test/java/org/codice/ddf/admin/application/service/command/ProfileListCommandTest.java]: ❗️ We should not be removing this command

**Comment by paouelle** [comment]
Should we also not update documentation with the removal of some commands?

File: platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 1 above)
Comment (tbatie) [reply platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: You could always put a profile as a boot feature, or define your own boot profile that points to the features you need. I'd prefer to completely decouple app from features if possible.

File: catalog/catalog-app/src/main/resources/features.xml [hunk 10]
```diff
@@ -76,7 +76,6 @@
 
     <feature name="catalog-core-plugins" install="auto" version="${project.version}"
```
[the remark this answers is not in the document]
Comment (tbatie) [reply catalog/catalog-app/src/main/resources/features.xml]: Yes, we could begin cleaning up feature files in general. I'd prefer to do that in a separate pr though, this one is already extremely large.

[the remark this answers is not in the document]
Comment (pklinef) [reply catalog/catalog-app/src/main/resources/features.xml]: Have you created that follow up ticket capturing what needs to be done yet?

File: platform/security/security-services-app/src/main/resources/features.xml (hunk 2 above)
Comment (pklinef) [reply platform/security/security-services-app/src/main/resources/features.xml]: ❗️ This still seems like the last hack that couples features to apps.  Ideally this should be removed in this PR if possible but a follow up ticket would also be acceptable if it will be too much.

File: distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json [hunk 11]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "admin-app",
```
Comment (pklinef) [inline distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json]: ❓ Do all of these really need `-app` in their names?  The same goes for the file name itself.  It all seems redundant since it is already in an `applications` folder.

**Comment by pklinef** [comment]
I was thinking the same thing about documentation.  `app-service-api.adoc` should be updated at least.

File: platform/security/security-services-app/src/main/resources/features.xml (hunk 2 above)
Comment (tbatie) [reply platform/security/security-services-app/src/main/resources/features.xml]: I thought this comment was left on the app.json files. I should be able to update the description of this feature in this pr without any additional changes :+1:

File: distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json (hunk 3 above)
Comment (tbatie) [reply distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json]: Although the backend no longer uses the end of the description, the frontend is parsing it to give it a display name. I originally removed them but the app names were no longer user friendly. I was going to add a displayName field and method to the Application interface until I started deprecating it.

File: catalog/catalog-app/src/main/resources/features.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (tbatie) [reply catalog/catalog-app/src/main/resources/features.xml]: I've created a follow up ticket for it https://codice.atlassian.net/browse/DDF-3645. This should have little to no impact by keeping these properties in for now.

File: platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationImpl.java [hunk 12]
```diff
@@ -40,93 +39,34 @@
...
-      version = mainFeature.getVersion();
-      description = mainFeature.getDescription();
-    } else {
-      Optional<Feature> first = features.stream().filter(f -> name.equals(f.getName())).findFirst();
-      if (first.isPresent()) {
-        mainFeature = first.get();
-        name = mainFeature.getName();
-        version = mainFeature.getVersion();
-        description = mainFeature.getDescription();
+  private List<String> bundleLocations;
```
[the remark this answers is not in the document]
Comment (tbatie) [reply platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationImpl.java]: Boon is the only one setting this field. I can create a default constructor.

File: distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json (hunk 11 above)
Comment (tbatie) [reply distribution/ddf-common/src/main/resources-filtered/etc/application-definitions/admin-app.json]: This names must remain identical or else the various plugins can't be associated with the app. I can change the file names though.

**Comment by tbatie** [comment]
@paouelle @pklinef There will be a separate pr for updating documentation, still wrapping it up now.

File: platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationImpl.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (paouelle) [reply platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationImpl.java]: just add a comment that it gets initialized via Boon bean mapping

[the remark this answers is not in the document]
Comment (paouelle) [reply platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationImpl.java]: By design, it means it could remain `null` if Boon doesn't see this attribute in the Json. You should add protection code against null above in `getBundles()`

File: platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBeanMBean.java [hunk 13]
```diff
@@ -20,87 +20,95 @@
...
-   * @return true if the application was successfully started, false if not.
+   * @return installation profile objects.
    */
-  boolean startApplication(String appName);
+  @Deprecated
+  List<Map<String, Object>> getInstallationProfiles();
 
   /**
-   * Stops an application with the given name.
+   * TODO:
```
Comment (figliold) [inline platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBeanMBean.java]: :question: What is the TODO?

File: platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml (hunk 1 above)
Comment (figliold) [reply platform/admin/core/admin-core-appservice/src/main/resources/OSGI-INF/blueprint/blueprint.xml]: :question: Did you actually remove https://github.com/codice/ddf/blob/master/distribution/ddf/src/main/filtered-resources/etc/org.codice.ddf.admin.applicationlist.properties?

File: catalog/solr/solr-app/src/main/resources/features.xml [hunk 14]
```diff
@@ -30,22 +30,22 @@
...
-
     <feature name="catalog-solr-solrclient" install="auto" version="${project.version}"
              description="Configure Solr parameters.">
         <feature prerequisite="true">catalog-solr-provider</feature>
         <bundle>mvn:ddf.catalog.solr/catalog-solr-solrclient/${project.version}</bundle>
     </feature>
 
+    <feature name="solr-app" install="auto" version="${project.version}"
+             description="The Solr Catalog Provider (SCP) is an implementation of the CatalogProvider interface using Apache Solr ${solr.version} as a data store.\nThe SCP supports extensible metacards, fast/simple contextual searching, indexes xml attributes/CDATA sections/XML text elements, contains full XPath support, works with an embedded local Solr Server, and also works with a standalone Solr Server.::Solr Catalog">
+        <feature prerequisite="true">catalog-app</feature>
```
Comment (figliold) [inline catalog/solr/solr-app/src/main/resources/features.xml]: :question: I noticed we remove prerequisite="true" in a lot of places.  Do we still need it here?

File: platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/Application.java [hunk 15]
```diff
@@ -18,39 +18,50 @@
 import org.apache.karaf.features.BundleInfo;
 import org.apache.karaf.features.Feature;
 
-/**
- * This class defines an application within DDF. An application is a component that can contain
- * multiple features. Inside each one of those features can be one or more bundles.
- */
+/** This class defines an application within DDF. An application is a collection of bundles. */
+@Deprecated
```
[the remark this answers is not in the document]
Comment (tbatie) [reply platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/Application.java]: Couldn't find any documentation on this. Going off stackoverflow and how intellij responds, deprecating a class doesn't necessarily mean the method is deprecated. I have no strong preference on this, but I'm leaving for now.

File: platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBeanMBean.java (hunk 13 above)
Comment (tbatie) [reply platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBeanMBean.java]: This is an existing TODO comment for the getServices method, none of the javadoc was changed in this class. Not sure why this is marked as an addition.

File: catalog/solr/solr-app/src/main/resources/features.xml (hunk 14 above)
Comment (tbatie) [reply catalog/solr/solr-app/src/main/resources/features.xml]: I only removed the ones that were causing cyclical dependencies. I'm concerned removing any additional ones will cause additional feature cleanup to happen in this PR.

**Comment by tbatie** [comment]
build now

**Comment by tbatie** [comment]
build now

**Comment by peterhuffer** [comment]
Successful hero. Started all apps and their configurations were the same as before the changes in this PR.
